### ISSUE TITLE
Security on redirect backend urls

### ISSUE THREAD

**Issue by 45kb** [hapijs/bell#190]
Hi, thanks for this awesome plugin, it works perfectly :+1: 

I just would like to ask you one thing that i do not quite understand.

I setted up the oauth backend urls for example `auth/facebook` where (after authorizing the app and got redirected) all the user infos are returned (the user profile).

Now, what i am wondering, is there any way that a malicious user/person could hack these urls like for example, sending fake user data inside query string / paypload / body of the request.

Is there any access_token/oauth_token verification by the Bell side before to return the user data on these routes?

To ask more shortly, is it possible to send fake user profile data on the Bell routes? :8ball: 

Thank you a lot!

**Comment by ldesplat** [comment]
4 things:
1. If your provider is on SSL which it has to be, and assuming NodeJS actually verifies the certificate chain, and you trust the root certificates that your O.S. has built in then you can be sure you're talking to a Facebook server and you don't have a man in the middle sending you bad data.
2. You also trust the DNS name resolution
3. https://github.com/hapijs/bell/blob/master/lib/providers/facebook.js#L24 is where we get the profile data. There we have to prove to Facebook we have the app secret and sign the auth token. You are only protected by SSL/Proper DNS resolution here.
4. Grabbing the auth/refresh tokens: https://github.com/hapijs/bell/blob/master/lib/oauth.js#L239

Now, do remember that this is not implicit auth as you would do when you do oauth2 login to facebook in client side javascript. You have much less opportunities to get hijacked, and therefore us going to the facebook /graph endpoint to verify the token would not help in any way.

Now if you do implicit auth through the facebook's sdk in client side js, when you send that token to your server, you better verify it & also exchange it for a long term token by signing it with your client secret. Bell does not help you in those scenarios.

**Comment by ldesplat** [comment]
We also as part of the oauth2 spec verify the state specified is the same as the one we set in the cookie during the process of getting a token as you're supposed to from the spec (spec mentions using localStorage potentially but most implementations use secure/httpOnly cookies). It is protected on top of that by us encrypting & signing that cookie so we know we set it and somebody did not fully hijack the user's session.

Anyways, closing this. Let me know if you have any more questions.
